SYCL/Basic/code_location_e2e.cpp

```diff
@@ -0,0 +1,141 @@
+// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
+
+/*
+    clang++ -fsycl -DNDEBUG -o smyl.bin code_location_e2e.cpp   // <<--- NDEBUG
+   will suppress fileLocation
+
+    -DNDEBUG  <-- not debugging, meaning fileName should not be retrievable (for
+   data privacy reasons)
+
+    This test uses hardcoded file code location positions. If modified, those
+   may change. pass in any value to signal "report only" and not run tests.
+
+*/
+
+// clang-format off
+//   we turn off clang-format, so that the line numbers don't get changed by reformatting.
+
+#include <sycl/sycl.hpp>
+using namespace sycl;
+
+bool stringsAreSameP(const char *a, const char *b){
+    // if both are nullptr, then they are the same. 
+    if ((a == nullptr) && (b == nullptr))
+      return true;
+    // but if only one, they are not
+    if ((a == nullptr) || (b == nullptr))
+      return false;
+
+    int index = 0;
+    while(true){
+        if(a[index] != b[index]) { return false; }
+        if(a[index] == '\0'){ return true; } // if we are on this line we know a[i]==b[i]
+        index++; 
+    }
+    // we will never arrive here
+    return true;
+}
+ 
+
+template <typename OS>
+void report(OS &out, detail::code_location code_loc){
+    out << "function {line:col} => " << code_loc.functionName()
+            << " {" << code_loc.lineNumber()
+            << ":" << code_loc.columnNumber()
+            << "}" << "\n";
+
+    auto fileName = code_loc.fileName();
+    if(fileName == nullptr)
+        out << "nullptr for fileName" << "\n";
+    else
+        out << "fileName: " << code_loc.fileName() << "\n";
+}
+
+template <typename OS>
+void test(OS &out, detail::code_location &code_loc,  const char  *fileName, 
+    const char *funcName,  int line, int col) {
+    
+  // functionName
+  auto funcNameStr = code_loc.functionName();
+  auto fNameResult = ((funcNameStr != nullptr) && stringsAreSameP(funcNameStr, funcName)  ) ? "OK" : "WRONG";
+  out << "code_location.functionName: " << fNameResult << "\n";
+
+  // lineNumber
+  auto lineNumberResult = (code_loc.lineNumber() == line) ? "OK" : "WRONG";
+  out << "code_location.lineNumber: " << lineNumberResult << "\n";
+
+  // columnNumber
+  auto colNumberResult = (code_loc.columnNumber() == col) ? "OK" : "WRONG";
+  out << "code_location.columnNumber: " << colNumberResult << "\n";
+
+  // fileName
+  auto fileNameStr = code_loc.fileName();
+#ifdef NDEBUG
+  //NDEBUG == not debugging == no fileName (for security)
+  auto fileNameResult = (fileNameStr == nullptr) ? "OK" : "WRONG - fileName should not be present when NDEBUG defined";
+#else
+   auto fileNameResult = stringsAreSameP(fileName, fileNameStr) ? "OK" : "WRONG";
+#endif
+  out << "code_location.fileName: " << fileNameResult << "\n";
+}
+
+int main(int argc, char** argv){
+    bool testing = (argc <= 1); // passing in ANY argument means report only, do not test
+    auto code_loc = sycl::detail::code_location::current();          // <--
```

Comment:
```suggestion
    auto code_loc = sycl::detail::code_location::current();          // <--
    int ExpectedLine = __LINE__ - 1;
```

then pass it to the `test` function.

Reply:
thanks!